File: week-7/py/member_system/routes/api_routes.py
from flask import Flask
from flask import request
from flask import session

from member_system.core import UnitOfWork

def configureAPIRoutes(app: Flask, unitOfWork: UnitOfWork):
    @app.route('/api/member', methods=['GET', 'PATCH'])
    def apiMember():
        if request.method == 'GET':
            return memberGet(unitOfWork)
        elif request.method == 'PATCH':
            return memberPatch(unitOfWork)

def memberGet(unitOfWork: UnitOfWork):
    username = request.args.get('username')
    isValid = username and session.get('name')
    if isValid:
        return memberGetIfValid(unitOfWork, username)
    return {'data': None}

def memberGetIfValid(unitOfWork: UnitOfWork, username: str):
    member = unitOfWork.memberRepository.getMemberByUsername(username)
    if member == None:
        return {'data': None}
    return { 
        'data': {
            'id': member.id,
            'name': member.name,
            'username': member.username
        }
    }

def memberPatch(unitOfWork: UnitOfWork):
    if request.content_type == 'application/json':
        id = session.get('id')
        newName = request.get_json()['name']
        if newName and id:
            return memberPatchIfValid(unitOfWork, id, newName)
    return {'error': True}

def memberPatchIfValid(unitOfWork: UnitOfWork, id: int, newName: str):
    success = unitOfWork.memberRepository.updateNameById(id, newName)
    if success:
        session['name'] = newName
        return {'ok': True}
    else:
        return {'error': True}

File: week-7/go/membersys/routes/api_routes.go
package routes

import (
	"github.com/gin-contrib/sessions"
	"github.com/gin-gonic/gin"
)

func (r *Router) memberGet(ctx *gin.Context) {
	username, success := ctx.GetQuery("username")
	session := sessions.Default(ctx)
	isValid := success && session.Get("name") != nil
	if isValid {
		r.memberGetIfValid(ctx, username)
		return
	}
	ctx.JSON(200, gin.H{"data": nil})
}

func (r *Router) memberGetIfValid(ctx *gin.Context, username string) {
	member := r.unitOfWork.MemberRepository.GetMemberByUsername(username)
	success := member.Id != -1
	if success {
		ctx.JSON(200, gin.H{
			"data": gin.H{
				"id":       member.Id,
				"name":     member.Name,
				"username": member.Username,
			},
		})
		return
	}
	ctx.JSON(200, gin.H{"data": nil})
}

func (r *Router) memberPatch(ctx *gin.Context) {
	if ctx.ContentType() == "application/json" {
		session := sessions.Default(ctx)
		id := session.Get("id")
		var nameRequestBody struct{ Name string }
		ctx.BindJSON(&nameRequestBody)
		if nameRequestBody.Name != "" && id != nil {
			r.memberPatchIfValid(ctx, id.(int), nameRequestBody.Name)
			return
		}
	}

	ctx.JSON(200, gin.H{"error": true})
}

func (r *Router) memberPatchIfValid(ctx *gin.Context, id int, newName string) {
	success := r.unitOfWork.MemberRepository.UpdateNameById(id, newName)
	if success {
		ctx.JSON(200, gin.H{"ok": true})
	} else {
		ctx.JSON(200, gin.H{"error": true})
	}
}
